teuthology/dispatcher/supervisor.py

```diff
@@ -151,6 +153,45 @@ def run_job(job_config, teuth_bin_path, archive_dir, verbose):
     if 'targets' in job_config:
         unlock_targets(job_config)
 
+def mark_machine_down(machine_name, count):
+    teuthology.lock.ops.update_lock(machine_name,
+                                    description='reimage failed {0} times'.format(count),
+                                    status='down',
+                                    )
+    log.exception(
+        'Reimage failed {0} times ... marking machine down'.format(count)
+    )
+
+def failure_is_reimage(failure_reason):
+    if not failure_reason:
+        return False
+    reimage_failure = "Error reimaging machines:"
+    if reimage_failure in failure_reason:
+        return True
+    else:
+        return False
+
+def check_for_failures_and_mark_down(targets, count=10):
+    # grab paddles history of jobs in the machine
+    # and count the number of reimaging Errors
+    # if it fails N times then mark the machine down
+    base_url = teuth_config.results_server
+    for k, _ in targets.items():
+        machine = k.split('@')[-1]
+        url = urljoin(
+                base_url,
+                '/nodes/{0}/jobs/?count={1}'.format(
+                machine, count)
+        )
+        resp = requests.get(url)
+        jobs = resp.json()
+        if len(jobs) < count:
+            return
+        for job in jobs:
+            if not failure_is_reimage(job['failure_reason']):
+                break
```

Comment:
This `break` will only step out of this inner loop, so healthy nodes would get reimaged too

Reply:
It might make more sense to do something like:
```
reimage_failures = list(filter(
    lambda j: failure_is_reimage(j['failure_reason']),
    jobs
))
if len(reimage_failures) < count:
    break
```